### PR TITLE
Feature/confirm auction

### PR DESCRIPTION
**Pull request by uv-orbs**

### CODE REVIEW

File: data/redisrepo/remove_order.go [hunk 1]
```diff
@@ -36,7 +36,9 @@ func (r *redisRepository) RemoveOrder(ctx context.Context, order models.Order) e
 
 	// update order status to CANCELED
 	orderIDKey := CreateOrderIDKey(order.Id)
-	transaction.HSet(ctx, orderIDKey, "status", models.STATUS_CANCELLED.String())
+	//transaction.HSet(ctx, orderIDKey, "status", models.STATUS_CANCELLED.String())
```
Comment (Luke-Rogerson) [inline data/redisrepo/remove_order.go:39]: Can you remove the comments here if we are now removing orders entirely when cancelled?

File: data/redisrepo/remove_order_test.go [hunk 2]
```diff
@@ -66,14 +66,14 @@ func TestRedisRepository_RemoveOrder(t *testing.T) {
 
 		buyPricesKey := CreateBuySidePricesKey(order.Symbol)
 		userOrdersKey := CreateUserOrdersKey(order.UserId)
-		orderIDKey := CreateOrderIDKey(order.Id)
+		//orderIDKey := CreateOrderIDKey(order.Id)
```
Comment (Luke-Rogerson) [inline data/redisrepo/remove_order_test.go:69]: Remove commented out code plz

File: data/redisrepo/store_order.go [hunk 3]
```diff
@@ -10,13 +10,9 @@ import (
 	"github.com/redis/go-redis/v9"
 )
 
-func (r *redisRepository) StoreOrder(ctx context.Context, order models.Order) error {
-
+func storeOrderTX(ctx context.Context, transaction redis.Pipeliner, order *models.Order) error {
```
Comment (Luke-Rogerson) [inline data/redisrepo/store_order.go:13]: Why did you feel the need to split this into a seperate function?

File: models/errors.go [hunk 4]
```diff
@@ -11,3 +11,4 @@ var ErrUnauthorized = errors.New("user not allowed to perform this action")
 var ErrOrderNotOpen = errors.New("order must be status open to perform this action")
 var ErrTransactionFailed = errors.New("transaction failed")
 var ErrInsufficientLiquity = errors.New("not enough liquidity in book to satisfy amountIn")
+var ErrAuctionInvalid = errors.New("orders in the auction can not fill any longer")
```
Comment (Luke-Rogerson) [inline models/errors.go:14]:
```suggestion
var ErrAuctionInvalid = errors.New("orders in this auction can no longer be filled")
```

File: service/auction.go [hunk 5]
```diff
@@ -0,0 +1,141 @@
...
+	s.orderBookStore.StoreOrders(ctx, res.Orders)
+
+	// add oredebook signature on the buffer
+	res.BookSignature = []byte("todo:sign")
+
+	return res, nil
+}
+
+func (s *Service) RevertAuction(ctx context.Context, auctionId uuid.UUID) error {
+	// get auction
```
Comment (Luke-Rogerson) [inline service/auction.go:90]: This is still todo?

File: service/auction.go [hunk 6]
```diff
@@ -0,0 +1,141 @@
...
+		return err
+	}
+	var filledOrders []*models.Order
+
+	// validate all pending orders fragments of auction
+	for _, frag := range frags {
+		// get order by ID
+		order, err := s.orderBookStore.FindOrderById(ctx, frag.OrderId, false)
+		if order == nil {
+			// cancel auction
```
Comment (Luke-Rogerson) [inline service/auction.go:113]: Put logs here please, particularly if it shouldn't happen

File: service/auction.go [hunk 7]
```diff
@@ -0,0 +1,141 @@
...
+			return models.ErrOrderNotFound
+		} else if !validatePendingFrag(frag, order) {
+			// cancel auction
+			s.orderBookStore.RemoveAuction(ctx, auctionId) // PANIC - shouldn't happen
+
+			logctx.Error(ctx, fmt.Sprintf("validatePendingFrag failed. PendingSize: %s FragSize:%s", order.SizePending.String(), frag.Size.String()))
+			return models.ErrAuctionInvalid
+		} else {
+			// fill fragment in the order
+			order.SizePending.Sub(frag.Size)
```
Comment (Luke-Rogerson) [inline service/auction.go:127]: Is this right? Shouldn't we add to ```SizePending```? And we know when an order has been fully filled when ```SizeFilled == Size```?

File: service/auction_test.go [hunk 8]
```diff
@@ -0,0 +1,69 @@
+package service_test
+
+// /////////////////////////////////////////////////////////////////
```
Comment (Luke-Rogerson) [inline service/auction_test.go:3]: Are you going to add tests for this?

File: transport/rest/auction.go [hunk 9]
```diff
@@ -0,0 +1,129 @@
...
+	w.WriteHeader(http.StatusOK)
+	_, err = w.Write(resp)
+	if err != nil {
+		logctx.Error(r.Context(), "failed to write confirmAuction response", logger.Error(err))
+		http.Error(w, "Error GetAmountOut", http.StatusInternalServerError)
+		return
+	}
+}
+
+func handleAuctionId(w http.ResponseWriter, r *http.Request) *uuid.UUID {
```
Comment (Luke-Rogerson) [inline transport/rest/auction.go:68]: I suggest making this as a middleware. You can see an example I did in `transport/rest/pagination.go`

File: transport/rest/server.go [hunk 10]
```diff
@@ -43,7 +43,10 @@ func (h *Handler) Listen() {
 	// LH side
 	lhApi := h.router.PathPrefix("/lh/v1").Subrouter()
 	lhApi.HandleFunc("/quote", h.amountOut).Methods("POST")
-	lhApi.HandleFunc("/approve_orders", h.approveOrders).Methods("GET")
+	lhApi.HandleFunc("/confirm_auction/{auctionId}", h.confirmAuction).Methods("GET")
```
Comment (Luke-Rogerson) [inline transport/rest/server.go:46]: None of these should be GET requests right? DELETE or POST/PUT?

File: service/auction.go (hunk 5 above)
Comment (uv-orbs) [reply service/auction.go:90]: yes

File: service/auction.go [hunk 11]
```diff
@@ -0,0 +1,141 @@
...
+	}
+	var filledOrders []*models.Order
+
+	// validate all pending orders fragments of auction
+	for _, frag := range frags {
+		// get order by ID
+		order, err := s.orderBookStore.FindOrderById(ctx, frag.OrderId, false)
+		if order == nil {
+			// cancel auction
+			s.orderBookStore.RemoveAuction(ctx, auctionId) // PANIC - shouldn't happen
```
[the remark this answers is not in the document]
Comment (uv-orbs) [reply service/auction.go:114]: still need to implement the panic here ...
lots of tests to do, lots of empty impls for now

File: transport/rest/server.go (hunk 10 above)
Comment (uv-orbs) [reply transport/rest/server.go:46]: yeah lets change that,

File: service/auction.go (hunk 11 above)
[the remark this answers is not in the document]
Comment (Luke-Rogerson) [reply service/auction.go:114]: Do you mean an actual [Go panic](https://gobyexample.com/panic)? If so, that's the wrong thing to be doing here as that will exit the entire server with a non-zero status. Better to handle gracefully and put lots of error logs so we have good visibility if this flow ever happens

**Pull request merged**